File: 백준/Gold/14719. 빗물/빗물.py
h, w = map(int, input().split())    # 세로, 가로 길이
block_height = list(map(int, input().split()))    # 블록 높이를 입력 받음
block = [[0] * w for _ in range(h)]

for i in range(w):
    bh = block_height[i]
    for j in range(bh):
        block[j][i] = 1

water = 0    # 물의 총량
for i in range(h):
    start, end = 0, 0
    for k in range(w):
        if block[i][k] == 1:
            start = k    # 왼쪽
            break
    for k in range(w-1, -1, -1):
        if block[i][k] == 1:
            end = k    # 오른쪽
            break

    if start == end:
        continue

    for k in range(start, end):
        if block[i][k] == 0:
            water += 1

print(water)

File: 백준/Gold/14719. 빗물/빗물.cc
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int h, w;
    cin >> h >> w;
    
    vector<int>block_height(w);
    for (int i=0; i<w; i++) {
        cin >> block_height[i];
    }
    
    vector<vector<int>> block(h, vector<int>(w, 0));
    
    for (int i=0; i<w; i++) {
        int bh = block_height[i];
        for (int j=0; j<bh; j++) {
            block[j][i] = 1;
        }
    }
    
    int water = 0;
    for (int i=0; i<h; i++) {
        int start = 0, end = 0;
        
        for (int k=0; k<w; k++) {    // 왼쪽
            if (block[i][k] == 1) {
                start = k;
                break;
            }
        }
        
        for (int k=w-1; k>-1; k--) {    // 오른쪽
            if (block[i][k] == 1) {
                end = k;
                break;
            }
        }
        
        if (start == end) {
            continue;
        }
        
        for (int k=start; k<end; k++) {
            if (block[i][k] == 0) {
                water++;
            }
        }
    }
    
    cout << water << endl;
    
    return 0;
}
